buildfile/src/main/scala/sbt/internal/Eval.scala

```diff
@@ -0,0 +1,450 @@
+
+  def bytes(s: String): Array[Byte] = s.getBytes("UTF-8")
+
+  /** The name of the synthetic val in the synthetic module that an expression is assigned to. */
+  private[sbt] final val WrapValName = "$sbtdef"
+
+  // used to map the position offset
+  class EvalSourceFile(name: String, startLine: Int, contents: String)
+      extends SourceFile(
+        new VirtualFile(name, contents.getBytes(StandardCharsets.UTF_8)),
+        contents.toArray
+      ):
+    override def lineToOffset(line: Int): Int = super.lineToOffset((line + startLine) max 0)
+    override def offsetToLine(offset: Int): Int = super.offsetToLine(offset) - startLine
+  end EvalSourceFile
+
+  trait EvalType[A]:
+    def makeSource(moduleName: String): SourceFile
+
+    /** Extracts additional information after the compilation unit is evaluated. */
+    def extract(run: Run, unit: CompilationUnit)(using ctx: Context): A
+
+    /** Deserializes the extra information for unchanged inputs from a cache file. */
+    def read(file: Path): A
+
+    /**
+     * Serializes the extra information to a cache file, where it can be `read` back if inputs
+     * haven't changed.
+     */
+    def write(value: A, file: Path): Unit
+
+    /** Extra information to include in the hash'd object name to help avoid collisions. */
+    def extraHash: String
+  end EvalType
+
+  class TypeExtractor extends tpd.TreeTraverser:
+    private[this] var result = ""
+    def getType(t: tpd.Tree)(using ctx: Context): String =
+      result = ""
+      this((), t)
+      result
+    override def traverse(tree: tpd.Tree)(using ctx: Context): Unit =
+      tree match
+        case tpd.DefDef(name, _, tpt, _) if name.toString == WrapValName =>
+          result = tpt.typeOpt.show
+        case t: tpd.Template   => this((), t.body)
+        case t: tpd.PackageDef => this((), t.stats)
+        case t: tpd.TypeDef    => this((), t.rhs)
+        case _                 => ()
+  end TypeExtractor
+
+  /**
+   * Tree traverser that obtains the names of vals in a top-level module whose type is a subtype of
+   * one of `types`.
+   */
+  class ValExtractor(tpes: Set[String]) extends tpd.TreeTraverser:
+    private[this] var vals = List[String]()
+
+    def getVals(t: tpd.Tree)(using ctx: Context): List[String] =
+      vals = Nil
+      traverse(t)
+      vals
+
+    def isAcceptableType(tpe: Types.Type)(using ctx: Context): Boolean =
+      tpe.baseClasses.exists { sym =>
+        tpes.contains(sym.fullName.toString)
+      }
+
+    def isTopLevelModule(sym: Symbols.Symbol)(using ctx: Context): Boolean =
+      (sym is Flags.Module) && (sym.owner is Flags.ModuleClass)
+
+    override def traverse(tree: tpd.Tree)(using ctx: Context): Unit =
+      tree match
+        case tpd.ValDef(name, tpt, _)
+            if isTopLevelModule(tree.symbol.owner) && isAcceptableType(tpt.tpe) =>
+          val str = name.mangledString
+          vals ::= (
+            if str.contains("$lzy") then str.take(str.indexOf("$"))
```

Comment:
In case literal identifiers are still supported in 2.x

For example, if I use `sbt new scala/hello-world.g8`:
 - If I first bump `build.properties` to use this 2.x SNAPSHOT
 - and if I then add the following to `build.sbt`: ```lazy val `hello-world` = (project in file("."))```

I then see an error: `java.lang.NoSuchMethodException: $Wrap65ae8c868c$.hello()`, and, digging a little, the val name gets mangled into `public sbt.Project $Wrap65ae8c868c$.hello$minusworld()` where a `$` char has been introduced.
```suggestion
            if str.contains("$lzy") then str.reverse.dropWhile(_ != '$').drop(1).reverse
```

Reply:
Thanks!